### PR TITLE
[JENKINS-53752] Add ability to exclude builds of untrusted PRs

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-53752](https://issues.jenkins-ci.org/browse/JENKINS-53752)

There's confusion around what "trusted" vs "untrusted" really means -
while ensuring we don't run the untrusted `Jenkinsfile`s is a good
start, there are cases where users may want to prevent any untrusted
pull request from being built at all. This adds an option to
`ForkPullRequestDiscoveryTrait` to filter out untrusted PRs completely.

(probably more testing is needed, and clarification in
`help-trust.html` may be needed too)

cc @stephenc @rsandell @dwnusbaum

### CODE REVIEW

**Comment by duncanmak** [comment]
@abayer This is great!

File: src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java [hunk 1]
```diff
@@ -67,6 +71,11 @@
             SCMRevision>
             trust;
 
+    /**
+     * TODO
+     */
+    private int untrustedHandling;
```
[the remark this answers is not in the document]
Comment (stephenc) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java]: ints were used in the others to simplify form binding but if someone can get enums to work on all paths i’d be happy to see

**Comment by abayer** [comment]
@dwnusbaum I couldn't figure out a way to test it properly without either live tests of some sort or really elaborate mocking, so I went with just manually testing.

@stephenc I've dealt with the `TODO`s, so I think this is good to go. Next step is up to you.

File: src/main/resources/org/jenkinsci/plugins/github_branch_source/Messages.properties [hunk 2]
```diff
@@ -10,6 +10,9 @@ ForkPullRequestDiscoveryTrait.headAndMerge=Both the current pull request revisio
 ForkPullRequestDiscoveryTrait.headOnly=The current pull request revision
 ForkPullRequestDiscoveryTrait.mergeOnly=Merging the pull request with the current target branch revision
 ForkPullRequestDiscoveryTrait.nobodyDisplayName=Nobody
+ForkPullRequestDiscoveryTrait.untrustedJenkinsfile=Untrusted pull requests' Jenkinsfiles will not be used
```
[the remark this answers is not in the document]
Comment (abayer) [reply src/main/resources/org/jenkinsci/plugins/github_branch_source/Messages.properties]: Nope, that's legit - I always forget to use escapes for apostrophes/quotes in `Messages.properties`. Fixed, and thanks for noticing it. =)

File: src/main/resources/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait/help-untrustedHandling.html [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<div>
+    Determines how untrusted pull requests are handled:
+    <dl>
+        <dt>Untrusted pull requests' <code>Jenkinsfile</code>s will not be used</dt>
+        <dd>Default behavior - the base <code>Jenkinsfile</code> will be used for untrusted pull requests, rather than
+            the <code>Jenkinsfile</code> from the untrusted pull request
+        </dd>
+        <dt>Untrusted pull requests will not be built</dt>
+        <dd>If a pull request is untrusted, it will not be built at all.
```
Comment (rsandell) [inline src/main/resources/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait/help-untrustedHandling.html]: Technically they won't even be discovered IIUC.

**Comment by rsandell** [comment]
Maybe a slightly better approach would be to still discover the PR but not automatically built it, like the default for tags? 
Someone can then go in manually and press build to get it verified. Or with some other magical plugin have it build when someone approves the PR.

there is also a faint echoing of one of the Jenkins 2 mantras "Secure by default" that maybe this behaviour should be set as default for newly created jobs? Since it is the more secure option?

**Comment by rsandell** [comment]
@abayer yea, that could work, though option ordering might need discussing?

1. Don't discover
2. Discover but don't (auto) build
3. Use trusted

**Comment by abayer** [comment]
@rsandell Actually, I think the "discover but don't auto-build" may be problematic - the job wouldn't get triggered initially, but wouldn't it get run as a result of webhooks when there are new commits, etc?

**Comment by abayer** [comment]
...and we'd have to make changes in `branch-api`, and `scm-api`, to add a new `SCMHead` type that we could then look for in `MultiBranchProject#isAutomaticBuild` (or maybe implement a `BranchBuildStrategy`?)...

Combining the webhooks issue and the complications in implementation for just disabling auto-build, I think we should table this option for now, at the very least not letting it block the specific feature that was requested. We can always add the additional option later.

And I'd also say that for now, we should keep the "default" behavior as the current behavior.

**Comment by stephenc** [comment]
I think this should be a simple choice: discover or don't discover untrusted. But I would just have it be a check-box on the strategies that are not "trust everything", e.g.

screenshot 2018-09-27 at 17 03 27

screenshot 2018-09-27 at 17 06 53

screenshot 2018-09-27 at 17 07 07

This could be easily achieved by having a common interface getter on the strategies for Fork discovery

**Comment by stephenc** [comment]
I think `ExcludeUntrustedPRsSCMHeadFilter` is an especially cumbersome way to implement this. The place where this makes sense is the place where the user is defining what is trusted. This means that the help for the field can actually have a meaningful description of what the effect of trust means.

Also might make sense for the Everyone option to show the check-box as disabled, e.g. 
screenshot 2018-09-27 at 17 10 53

So that people selecting that understand that trusting everyone means nobody is untrusted and thus all fork PRs will always be discovered too

**Comment by abayer** [comment]
Ok, lemme think on a better way to implement this. Thanks!

**Comment by abayer** [comment]
...though from what I can remember, an `SCMHeadFilter` was the only way I could implement "don't discover".

**Comment by stephenc** [comment]
I'd probably add a flag to the context and then have the forkpr strategy control that flag

**Comment by abayer** [comment]
@stephenc Hrm. I'll see what I can do.

**Comment by brianjmurrell** [comment]
Has this been abandoned, back-burnered or something else?  It's a much needed requirement to satisfy security concerns of random folks forking your repo and pushing PRs.

**Comment by brianjmurrell** [comment]
Is any further work on this planned or has it been abandoned and should be closed?

**Comment by stephenc** [comment]
@brianjmurrell so IMHO the correct thing to do is use a branch build strategy that doesn't auto-build untrusted PRs. That way you can manually trigger a build of that one specific untrusted PR that you have just reviewed.

This PR is attempting to not even discover untrusted PRs which would mean that they do not even appear in the Jenkins UI (leading to questions like: "Where is the Jenkins job for PR-256?"

It is relatively trivial to create a plugin that is based on https://github.com/jenkinsci/basic-branch-build-strategies-plugin/blob/master/src/main/java/jenkins/branch/buildstrategies/basic/ChangeRequestBuildStrategyImpl.java#L69-L83 but just additionally checks if the revision is trusted (you have the SCMSource). If this is a critical requirement for you, that is the approach I recommend (I think in the case of trust, I would accept a PR against basic branch build strategies that adds an "Only change requests from trusted origins" checkbox... though it is slightly more advanced than I'm trying to keep basic)

**Comment by stephenc** [comment]
@abayer said:

> Actually, I think the "discover but don't auto-build" may be problematic - the job wouldn't get triggered initially, but wouldn't it get run as a result of webhooks when there are new commits, etc?

Actually if you use a `BranchBuildStrategy` then webhooks *should* be ignored too as the event is processed by branch-api which will consult the `BranchBuildStrategy` to decide if the branch should be built.

Discovery might be better handled by a filter (that would work for everyone and could just be part of scm-api) that checks if the SCMHead is trusted.

**Comment by brianjmurrell** [comment]
> so IMHO the correct thing to do is use a branch build strategy that doesn't auto-build untrusted PRs.

I'm on board with that.

> That way you can manually trigger a build of that one specific untrusted PR that you have just reviewed.

As in I could go to https://jenkins.example.com/job/<github_org>/job/<github_repo>/view/change-requests/ to find a PR that was not auto-built because it was from an untrusted source, click on it and then initiate a "Build Now"?

If so, that would be sweet.

> This PR is attempting to not even discover untrusted PRs which would mean that they do not even appear in the Jenkins UI (leading to questions like: "Where is the Jenkins job for PR-256?"

Yeah.  I like that less than your proposal above, now that you have proposed it.

**Comment by duncanmak** [comment]
@abayer ping? What do you think about the state of this PR?

**Comment by stephenc** [comment]
@abayer is this required now that https://github.com/jenkinsci/basic-branch-build-strategies-plugin/commit/75c899763df1aca8658717178bc742403455419a is available?

Screenshot 2019-03-21 at 15 19 43

**Comment by stephenc** [comment]
*NOTE TO OTHERS*

This is probably the starting configuration you want if you have the potential of PRs from untrusted people... tweak from there as necessary

Screenshot 2019-03-21 at 15 21 25

**Comment by conn** [comment]
It would be cool if the help text were more clear in the context of GitHub. The wording seems to be vague in order to support all branch sources.

**Comment by stephenc** [comment]
@conn the world does not revolve around github (even if it can seem that way sometimes)

**Comment by brianjmurrell** [comment]
Not to sound ungrateful for these new features, but is/will it be possible to manually select an untrusted PR that has been prevented from automatically building to be built, say, in the case that somebody trusted inspects the PR and decides that it can go ahead and be built?

**Comment by stephenc** [comment]
@brianjmurrell yes exactly the is possible as the PRs are discovered. You just request a build in the Jenkins UI

**Comment by brianjmurrell** [comment]
Super nice!  And so that's available in the [_Basic Branch Build Strategies Plugin_ version 1.2.0](https://github.com/jenkinsci/basic-branch-build-strategies-plugin/releases/tag/basic-branch-build-strategies-1.2.0) that was released a few hours ago?

If so, how long does that take to hit the Jenkins plugin repo?

**Comment by stephenc** [comment]
Already there. The screenshot is from my own instance updated from the OSS update centre

**Comment by conn** [comment]
@stephenc I got confused at the wording. With the help text for `ForkPullRequestDiscoveryTrait` it makes total sense. Great work! Can't wait to try it.

**Comment by bitwiseman** [comment]
@abayer @stephenc 
So, this PR can be closed along with related JIRA?

**Comment by bitwiseman** [comment]
@abayer - I'm going to close this PR. Please reopen if I'm mistaken.

**Comment by brianjmurrell** [comment]
I would disagree with closing this.  I have filed [JENKINS-58618](https://issues.jenkins-ci.org/browse/JENKINS-58618) and [JENKINS-58683](https://issues.jenkins-ci.org/browse/JENKINS-58683) about defects in this change, both of which have had not even been triaged in https://issues.jenkins-ci.org/.

**Comment by bitwiseman** [comment]
@brianjmurrell
Would you like to take ownership of this PR?  You are totally welcome to open a PR of your own with your own proposed set of changes.

That said, I'm a little confused. The two issues you filed are against the basic branch build strategies plugin so the fix to those changes probably wouldn't be made in this repo.

**Pull request closed without merging**